src/python/pants/backend/java/package/deploy_jar.py

```diff
@@ -0,0 +1,165 @@
+)
+from pants.core.util_rules.archive import ZipBinary
+from pants.engine.addresses import Addresses
+from pants.engine.fs import AddPrefix, CreateDigest, Digest, FileContent, MergeDigests
+from pants.engine.process import BashBinary, Process, ProcessResult
+from pants.engine.rules import Get, collect_rules, rule
+from pants.engine.target import Dependencies, DependenciesRequest
+from pants.engine.unions import UnionRule
+
+logger = logging.getLogger(__name__)
+
+
+_JAVA_MANIFEST_FILENAME = "META-INF/MANIFEST.MF"
+_PANTS_MANIFEST_PARTIAL_JAR_FILENAME = "pants_manifest_only.notajar"
+_PANTS_CAT_AND_REPAIR_ZIP_FILENAME = "_cat_and_repair_zip_files.sh"
+
+
+@dataclass(frozen=True)
+class DeployJarFieldSet(PackageFieldSet):
+    required_fields = (
+        JvmMainClassName,
+        Dependencies,
+    )
+
+    main_class: JvmMainClassName
+    output_path: OutputPathField
+    dependencies: Dependencies
+
+
+@rule
+async def package_deploy_jar(
+    bash: BashBinary,
+    zip: ZipBinary,
+    field_set: DeployJarFieldSet,
+) -> BuiltPackage:
+    """
+    Constructs a deploy ("fat") JAR file (currently from Java sources only) by
+    1. Resolving/compiling a classpath for the `root_address` target,
+    2. Producing a ZIP file containing _only_ the JAR manifest file for the `main_class`
+    3. Creating a deploy jar with a broken ZIP index by concatenating all dependency JARs together,
+       followed by the thin JAR we created
+    4. Using the unix `zip` utility's repair function to fix the broken fat jar
+    """
+
+    if field_set.main_class.value is None:
+        raise Exception("Needs a `main` argument")
+
+    #
+    # 1. Produce a thin JAR containing our first-party sources and other runtime dependencies
+    #
+
+    dependencies = await Get(Addresses, DependenciesRequest(field_set.dependencies))
+    classpath = await Get(Classpath, Addresses, dependencies)
+
+    #
+    # 2. Produce JAR manifest, and output to a ZIP file that can be included with the JARs
+    #
+
+    main_class = field_set.main_class.value
+
+    manifest_content = FileContent(
+        _JAVA_MANIFEST_FILENAME,
+        # NB: we're joining strings with newlines, becuase the JAR manfiest format
+        # needs precise indentation, and _cannot_ start with a blank line. `dedent` seriously
+        # messes up those requirements.
+        "\n".join(
+            [
+                "Manifest-Version: 1.0",
+                f"Main-Class: {main_class}",
+                "",  # THIS BLANK LINE WILL BREAK EVERYTHING IF DELETED. DON'T DELETE IT.
+            ]
+        ).encode("utf-8"),
+    )
+
+    manifest_jar_input_digest = await Get(Digest, CreateDigest([manifest_content]))
+    manifest_jar_result = await Get(
+        ProcessResult,
+        Process(
+            argv=[
+                zip.path,
+                _PANTS_MANIFEST_PARTIAL_JAR_FILENAME,
+                _JAVA_MANIFEST_FILENAME,
+            ],
+            description="Build thin JAR file.",
```

Comment:
This could be more descriptive, e.g. `Build manifest JAR file for ADDRESS`, but shouldn't block landing this PR.